### PR TITLE
Added Deployment Time

### PR DESCRIPTION
**Pull request by rafaelraposospot**
# Hey, I just made a Pull Request!

## Description

Adding deploymentTime to WorkflowConfiguration when a Workflow is created.

## Motivation and Context

It makes is easy for a user to check when a workflow was created

## Checklist for PR author(s)

- [ ] Changes are covered by unit test
- [ ] All tests pass
- [ ] Code coverage check passes
- [ ] Error handling is tested
- [ ] Errors are handled at the appropriate layer
- [ ] Errors that cannot be handled where they occur are propagated
- [ ] (optional) Changes are covered by system test
- [ ] Relevant documentation updated
- [ ] This PR has NO breaking change to public API
- [ ] This PR has breaking change to public API and it is documented

## Checklist for PR reviewer(s)

- [ ] This PR has been incorporated in release note for the coming version
- [ ] Risky changes introduced by this PR have been all considered

### CODE REVIEW

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 1]
```diff
@@ -103,8 +119,9 @@ public WorkflowResource(Storage storage, WorkflowValidator workflowValidator,
             Middlewares.<Workflow>authed(requestAuthenticator).and(json()), "POST", BASE + "/<cid>",
             rc -> ac -> createOrUpdateWorkflow(arg("cid", rc), rc, ac)),
         Route.with(
-            Middlewares.<ByteString>authed(requestAuthenticator).and(json()), "DELETE", BASE + "/<cid>/<wfid>",
-            rc -> ac -> deleteWorkflow(arg("cid", rc),arg("wfid", rc), ac)),
+            Middlewares.<ByteString>authed(requestAuthenticator).and(json()), "DELETE",
```
Comment (sonjaer) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java:122]: Could we do formatting in another PR to avoid cluttering this one?

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 2]
```diff
@@ -103,8 +119,9 @@ public WorkflowResource(Storage storage, WorkflowValidator workflowValidator,
...
+            BASE + "/<cid>/<wfid>",
+            rc -> ac -> deleteWorkflow(arg("cid", rc), arg("wfid", rc), ac)),
```
Comment (ckiosidis) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java:124]: try to avoid refactoring in commits that add/change the logic. It makes it more difficult to review the actual changes

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 3]
```diff
@@ -151,10 +165,13 @@ public WorkflowResource(Storage storage, WorkflowValidator workflowValidator,
     if (payload.isEmpty()) {
       return Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Missing payload."));
     }
-    final WorkflowConfiguration workflowConfig;
+    WorkflowConfiguration workflowConfig;
     try {
       workflowConfig = OBJECT_MAPPER
           .readValue(payload.get().toByteArray(), WorkflowConfiguration.class);
+      workflowConfig = WorkflowConfigurationBuilder.from(workflowConfig).deploymentTime(clock.instant()).build();
+
```
Comment (ckiosidis) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java:173]: i am not entirely sure if this is the best place to do this because I worry that we are might miss different entrypoints for creating/updating workflows.
Or we could extract these two lines in a different function `readFromJsonWithDefaults`

Comment (ckiosidis) [reply styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java:173]: or createWorkflowFromJsonWithDefaults

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java [hunk 4]
```diff
@@ -113,8 +114,10 @@
   @Mock private WorkflowActionAuthorizer workflowActionAuthorizer;
   @Mock private GoogleIdToken idToken;
   @Mock private RequestAuthenticator requestAuthenticator;
+  @Mock private Clock clock;
 
   private static final String SERVICE_ACCOUNT = "foo@bar.iam.gserviceaccount.com";
+  private static final Instant deploymentTime = Instant.ofEpochSecond(0);
```
Comment (ckiosidis) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java:120]: maybe use another time/date here?

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java [hunk 5]
```diff
@@ -184,7 +188,7 @@ protected void init(Environment environment) {
     when(requestAuthenticator.authenticate(any())).thenReturn(() -> Optional.of(idToken));
     WorkflowResource workflowResource =
         new WorkflowResource(storage, workflowValidator, workflowInitializer, workflowConsumer,
-            workflowActionAuthorizer);
+            workflowActionAuthorizer, clock);
```
Comment (narape) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java:191]: @rafaelraposospot no need to mock the clock. You can pass a fixed [clock](https://docs.oracle.com/javase/8/docs/api/java/time/Clock.html#fixed-java.time.Instant-java.time.ZoneId-)
```suggestion
            workflowActionAuthorizer, Clock.fixed(deploymentTime, ZoneOffset.UTC);
```

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java [hunk 6]
```diff
@@ -113,8 +114,10 @@
   @Mock private WorkflowActionAuthorizer workflowActionAuthorizer;
   @Mock private GoogleIdToken idToken;
   @Mock private RequestAuthenticator requestAuthenticator;
+  @Mock private Clock clock;
```
Comment (narape) [inline styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java:117]:
```suggestion
```
See my other comment, no need to mock the clock

File: styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java (hunk 4 above)
Comment (narape) [reply styx-api-service/src/test/java/com/spotify/styx/api/WorkflowResourceTest.java:120]: The thing is that the epoc is the more probable value to be used as default value so using something else is better. You can use even Instant.now()

File: styx-common/src/main/java/com/spotify/styx/testdata/TestData.java [hunk 7]
```diff
@@ -172,6 +173,7 @@
           .runningTimeout(Duration.parse("PT20H"))
           .retryCondition("#tries<2")
           .env("foo","bar")
+          .deploymentTime(Instant.ofEpochSecond(0))
```
Comment (narape) [inline styx-common/src/main/java/com/spotify/styx/testdata/TestData.java:176]:
```suggestion
          .deploymentTime(WORKFLOW_DEPLOYMENT_TIME)
```
Maybe you can create a constant and use it on other test too

**Review by narape — APPROVED** [review]

File: styx-e2e-test/src/test/java/com/spotify/styx/e2e_tests/EndToEndTestBase.java [hunk 8]
```diff
@@ -186,6 +189,7 @@ private void startStyx() {
     styxApiThread = executor.submit(() -> {
       HttpService.boot(env -> StyxApi.newBuilder()
           .setServiceName(API_SERVICE_NAME)
+          .setTime(time)
```
Comment (ckiosidis) [inline styx-e2e-test/src/test/java/com/spotify/styx/e2e_tests/EndToEndTestBase.java:192]: nice

**Review by ckiosidis — APPROVED** [review]
